### PR TITLE
DAT-55474 Timbermill 2 - persistence - cron

### PR DESCRIPTION
**Pull request by oz-zafar**

### CODE REVIEW

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 1]
```diff
@@ -86,6 +86,10 @@
 	private int maxBulkIndexFetches; // after such number of fetches, bulk is considered as failed and won't be persisted anymore
 	private LinkedBlockingQueue<Pair<DbBulkRequest, Integer>> failedRequests = new LinkedBlockingQueue<>(100000);
 	private DiskHandler diskHandler;
+	private int numOfBulksPersistedToDisk = 0;
+	private int numOfSuccessfullBulksFromDisk = 0;
+	private int numOfFetchedMaxTimes = 0;
+	private int numOfCouldntBeInserted = 0;
```
Comment (oz-zafar) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:92]: all these members are only temporary - they will help to analyze the persistence performance

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/PersistentFetchJob.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import com.datorama.oss.timbermill.ElasticsearchClient;
+import com.datorama.oss.timbermill.common.ElasticsearchUtil;
+
+public class PersistentFetchJob implements Job {
+
+	private static final Logger LOG = LoggerFactory.getLogger(PersistentFetchJob.class);
+	private static boolean keepRunning = false;
+
+	@Override public void execute(JobExecutionContext context) {
+		if (!keepRunning) {
```
Comment (oz-zafar) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/PersistentFetchJob.java:17]: keepRuning will stop jobs to be executed while another job os successfully sending bulks to es. not sure if we need a lock here?

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java [hunk 3]
```diff
@@ -704,5 +710,30 @@ private DbBulkRequest extractFailedRequestsFromBulk(DbBulkRequest dbBulkRequest,
...
+	public boolean retryFailedRequestsFromDisk() {
+		boolean keepRunning = false;
+		if (withPersistence && diskHandler.hasFailedBulks()){
+			keepRunning = true;
+			LOG.info("------------------ Failed Requests From Disk Retry Start ------------------");
+			List<DbBulkRequest> failedRequestsFromDisk = diskHandler.fetchAndDeleteFailedBulks();
+			for (DbBulkRequest dbBulkRequest : failedRequestsFromDisk) {
+				if (!sendDbBulkRequest(dbBulkRequest, 0)){
+					keepRunning = false;
+				}
```
Comment (oz-zafar) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:726]: if a job fails to retry a request then it won't keep fetching bulks from disk so there will be a break until the next job will be executed.
Two points:
1.  in the code above we take a break after one failed bulks. do u think we should choose a higher limit?
2. in some cases, this break can be very short (a short time after the break, 10 minutes passed from the last job and a new one is executed). we can sleep for some time so the next job will wait

Comment (edengal) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/ElasticsearchClient.java:726]: 1. I think one fail is good enough
2. No need for a sleep

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/PersistentFetchJob.java (hunk 2 above)
Comment (edengal) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/PersistentFetchJob.java:17]: OK but let's change it a bit to make it more understandable:
* change keepRunning name to currentlyRunning
* set it to false once you finish running
* retryFailedRequestsFromDisk souldn't be assigned, just keep running until the return value is false.

File: timbermill-server/src/main/java/com/datorama/timbermill/server/service/TimbermillService.java [hunk 4]
```diff
@@ -59,6 +59,7 @@ public TimbermillService(@Value("${INDEX_BULK_SIZE:200000}") Integer indexBulkSi
 			@Value("${CACHE_MAX_HOLD_TIME_MINUTES:6}") int maximumCacheMinutesHold,
 			@Value("${WITH_PERSISTENCE:true}") boolean withPersistence,
 			@Value("${DISK_HANDLER_STRATEGY:sqlite}") String diskHandlerStrategy,
+			@Value("${persistent.fetch.cron.expression:0 0/10 * 1/1 * ? *}") String persistentFetchCronExp,
```
Comment (edengal) [inline timbermill-server/src/main/java/com/datorama/timbermill/server/service/TimbermillService.java:62]: use upper case and undescores

**Pull request merged**